MPI Read in Batches

c-bebop opened the pull request:
This enables an application to read in batches of edges in parallel using our MPI Graph I/O code.

domischweisgut commented on include/gdsb/batcher.h:
```diff
@@ -83,8 +83,41 @@ template <class EdgeContainer> class Batcher
+{
+    return edge_count / expected_batch_size;
+}
+
```
maybe you want to make clear that you use expected_batch_size? i.e. count_of_batches_expected? but I also think that it is clear when you read the arguments, so I don't know.

c-bebop replied:
Yes, in this case I wanted to let the parameter names _speak for themselves_.

domischweisgut commented on include/gdsb/batcher.h:
```diff
@@ -83,8 +83,41 @@ template <class EdgeContainer> class Batcher
+    uint64_t const batch_count = edge_count / max_batch_size;
+    uint64_t const remaining = edge_count % max_batch_size;
+    uint64_t fair_size = max_batch_size + (remaining / batch_count);
+    return fair_size;
```
Here fair_size just gets bigger. Additionally, by the behavior of integer division, you will probably only add a few edges?

c-bebop replied:
That's intended and only used for tests.

domischweisgut commented on include/gdsb/mpi_graph_io.h:
```diff
@@ -132,6 +132,33 @@ std::tuple<Vertex64, uint64_t> all_read_binary_graph_partition(MPI_File const in
+    }
+
+    MPI_Status status;
+    int const read_all_error = MPI_File_read_all(input, edges, count, mpi_datatype, &status);
```
When you call MPI_File_read_all you try to read 'count' elements (the batch size I guess?). I think what happens if the end of the file (EOF) comes before it reads all elements as specified by counter, is the following: MPI reads the elements until EOF and writes them to the buffer (but I think you have to get the count before EOF to make sure that you don't read bullshit from the buffer). But I think the return value won't be MPI_SUCCESS. So I think you have to make more distinctions for the error_message maybe to make sure the read process does not crash unnecessarily?

c-bebop replied:
For now, we will not guarantee that we won't read beyond EOF. OpenMPI or MPICH does not provide enough information on the actual field in `status` used to write the count of read elements. I've left some comments in code and also documentation of the function to address the issue.

c-bebop replied:
Update: see comment on this issue.

domischweisgut commented on test/mpi_graph_io_tests.cpp:
```diff
@@ -622,4 +622,373 @@ TEST_CASE("MPI", "read")
+    REQUIRE(edges.size() == count_15);
+    mpi::all_read_binary_graph_batch(binary_graph.get(), header, &(edges[0]), sizeof(Edge32), 0u, count_15, mpi_edge_t.get());
+    idx = 0;
+
```
Here in this check you avoid the 'problem' in  all_read_binary_graph_batch by setting the count precisely when you read the last batch. If you do that every time when you read a graph this is fine. But if not (if you use a simple loop for example), you should probably rewrite the test to see how all_read_binary_graph_batch behaves.

domischweisgut reviewed (approved):
If you read the files as in the test, everything is fine. If not, I think you have to look at the batch read function for MPI again.

c-bebop commented:
> If you read the files as in the test, everything is fine. If not, I think you have to look at the batch read function for MPI again.

I've addressed the issue of potentially reading past EOF by tracking the count of read in edges along with parameters for batch size and edge size in bytes using `ReadBatch` struct. Doing so will guarantee that we never read more edge than actually present in the binary file.

Now, instead of living with the limitations of the MPI implementations, we offer an API that is safe from reading past EOF. I've also removed the offset functionality since it's not used nor tested.
